pkg/selector/pod/legacy.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2022 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package pod
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/mock"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector/generic"
+	genericnamespace "github.com/chaos-mesh/chaos-mesh/pkg/selector/generic/namespace"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector/generic/registry"
+)
+
+// LegacyPodSelector is the temporary place to hold the legacy logic for selecting then filtering the Pod.
+type LegacyPodSelector struct {
```

Comment:
Why not merge `LegacyPodSelector` and `SelectImpl` into a structure? As this `LegacyPodSelector` is also a breaking change to package `pkg/selector/pod`.

Reply:
Do you mean embed `LegacyPodSelector` in to `SelectImplSelectImpl`? Or only move all the methods of `LegacyPodSelector` into methods of `SelectImpl`?

The former one makes not soo many changes, we could make it. 

But for the latter one, it would make more trouble when we want to purge `LegacyPodSelector` in the future. Actually, one of my main purposes is to tidy the usage of `LegacyPodSelector`, for purging it in the future.

Reply:
I think that bringing breaking changes on `pkg/selector/pod` is acceptable, unlike kubernetes CRD API, we do not make any promise on `pkg/selector/pod` (and either most of other packages ), we actually do not intend other users use it as a library.

Reply:
Yes, breaking changes don't matter. I prefer to embed `LegacyPodSelector` into `SelectImpl`, then we can ignore the existence of `LegacyPodSelector` (only construct `SelectImpl`).

Reply:
> Yes, breaking changes don't matter. I prefer to embed `LegacyPodSelector` into `SelectImpl`, then we can ignore the existence of `LegacyPodSelector` (only construct `SelectImpl`).

In this solution, we should privatize the `NewLegacyPodSelector`.

Reply:
🤔 But it would make the original dependency on `LegacyPodSelector` "extended" to dependency on `pod.SelectImpl`, which bring 2 problems:

- it would use `selector pod.SelectImpl` as `select.SelectPods` **directly**, which might break the convention to using   `Selector`(in `pkg/selector/selector.go`) and each `SelecImpl` in each package, beacse it seems `pkg/selector` is designed to using `Selector`(in `pkg/selector/selector.go`) instead of designed to using `SelectImpl` directly. Or maybe I miss something? PTAL @iguoyr @YangKeao 
- the current constructor of `pod.SelectImpl` is coupled with `fx` dependency injection library.

Reply:
I think the `fx` dependency injection doesn't prevent us from constructing an object manually, especially outside of controller-manager (like the apiserver). For applications embedded in controller-manager, we can also make them by `fx` (we will construct ctrlserver by `fx` in #2992).

Reply:
Oh, that's right, the issue about `fx` could be resolved. What do you think about the former one?

Reply:
> It seems `pkg/selector` is designed to using `Selector`(in `pkg/selector/selector.go`) instead of designed to using `SelectImpl` directly. Or maybe I miss something? PTAL @iguoyr @YangKeao

I don't think so, as we expose `pod.SelectImpl` as a public type, and provide it by `fx`, we allow other components to use `pod.SelectImpl` straightly (and its methods are common enough to be used). If we only allow other components to use `Selector`, we should make `pod.SelectImpl` private and only expose an interface.

Reply:
>  If we only allow other components to use Selector, we should make pod.SelectImpl private and only expose an interface.

The current codes indeed expose these `SelectImpl` structs, but I think that is not the original intent...  🙈🙈🙈

Reply:
I think it's fine to use `SelectImpl` directly or not. But if we do not want to use `SelectImpl` directly, maybe we should not embed `pod.SelectImpl` into `container.SelectImpl` at the same time (as they are under two different packages)? 🤔 

https://github.com/chaos-mesh/chaos-mesh/blob/d77e86e6d75d585815f4bf8a95484ca5e612f0e6/pkg/selector/container/selector.go#L31-L37

Reply:
> I think it's fine to use `SelectImpl` directly or not. But if we do not want to use `SelectImpl` directly, maybe we should not embed `pod.SelectImpl` into `container.SelectImpl` at the same time (as they are under two different packages)? 🤔
> 
> https://github.com/chaos-mesh/chaos-mesh/blob/d77e86e6d75d585815f4bf8a95484ca5e612f0e6/pkg/selector/container/selector.go#L31-L37

Why not merge package `selector/pod` into `selector/container`? As the `selector/container` strongly depend on `selector/pod`.

Reply:
> But if we do not want to use SelectImpl directly, maybe we should not embed pod.SelectImpl into container.SelectImpl at the same time (as they are under two different packages)?

Please do not be restricted by golang's syntax, golang does not provide access levels other than `exported` or `unexported`. Exporting the struct but only for internal usage still commonly exists. 

I want to figure out whether `SelectImpl` is recommended to use without the limitation of  who is the invoker.

Reply:
> Why not merge package selector/pod into selector/container? As the selector/container strongly depend on selector/pod.

I even prefer to flat the whole `pkg/selector/*` into one layer of the package. 🤣

Reply:
> I even prefer to flat the whole `pkg/selector/*` into one layer of the package. 🤣

Good idea😏

Reply:
> Exporting the struct but only for internal usage still commonly exists.

Yes, but I think it's not a best practice. I agree with @Hexilee, if still using `SelectImpl`, that means it could be invoked by outside packages. The struct `Selector` helps to reflect the detail `xxxSelector`, but it doesn't prevent to invoke `xxxSelector` directly. And if we don't want to invoke directly, maybe move `pkg/selector/container` package or create a legacyPodSelector when `Select` in `pkg/selector/container` are better choices.

What do you think? @YangKeao